sotodlib/tod_ops/sss.py

```diff
@@ -0,0 +1,216 @@
+import numpy as np
+from numpy.polynomial import legendre as L
+from scipy.optimize import curve_fit
+from scipy.interpolate import interp1d
+from sotodlib import core, tod_ops
+from sotodlib.tod_ops import bin_signal
+import logging
+
+logger = logging.getLogger(__name__)
+
+def bin_by_az(aman, signal=None, az=None, bin_range=[-np.pi, np.pi], bins=3600, flags=None):
+    """
+    Bins a signal by azimuth angle.
+
+    Parameters
+    ----------
+    aman: TOD
+        core.AxisManager
+    signal: array-like, optional
+        numpy array of signal to be binned. If None, the signal is taken from aman.signal.
+    az: array-like, optional
+        A 1D numpy array representing the azimuth angles. If not provided, the azimuth angles are taken from aman.boresight.az attribute.
+    bin_range: array-like, optional
+        A list specifying the range of azimuth angles to consider for binning. Defaults to [-np.pi, np.pi].
+        If None, [min(az), max(az)] will be used for binning.
+    bins: integer
+        The number of bins to divide the azimuth angle range into. Defaults to 360.
+    flags: RangesMatrix, optional
+        Flag indicating whether to exclude flagged samples when binning the signal.
+        Default is no mask applied.
+
+    Returns
+    -------
+    binning_dict: A dictionary containing the binned signal values.
+        * 'bin_centers': center of each bin of azimuth
+        * 'binned_signal': binned signal
+        * 'binned_signal_sigma': estimated sigma of binned signal
+    """
+    if signal is None:
+        signal = aman.signal
+    if az is None:
+        az = aman.boresight.az
+    binning_dict = bin_signal(aman, bin_by=az, signal=signal,
+                               bin_range=bin_range, bins=bins, flags=flags)
+    return binning_dict
+
+def fit_sss(az, sss_stats, nmodes, fit_range=None):
+    """
+    Function for fitting legendre polynomials to signal binned in azimuth.
+
+    Paramaters
+    ----------
+    az: array-like
+        azimuth array from main axis manager
+    sss_stats: AxisManager
+        Axis manager containing binned signal and azimuth used for fitting.
+        Created by ``get_sss`` function.
+    nmodes: integer, required
+        Highest order legendre polynomial to include in the fit.
+    fit_range: list
+        Azimuth range used to renormalized to the [-1,1] range spanned
+        by the legendre polynomials for fitting. Default is the max-min
+        span in the ``binned_az`` array passed in via ``sss_stats``.
+
+    Returns
+    -------
+    sss_stats: AxisManager
+        Returns updated sss_stats with added fit information
+    model_sig_tod: array-like
+        Model fit for each detector size ndets x n_samps
+    """
+    bin_width = sss_stats.binned_az[1] - sss_stats.binned_az[0]
+    m = ~np.isnan(sss_stats.binned_signal[0]) # masks bins without counts
+    if np.count_nonzero(m) < nmodes:
```

Comment:
This is check is inaccurate if nmodes is the _order of the highest mode_.  You should fail on count(m) < nmodes + 1 (unless you change change the meaning of nmodes as discussed above).

Reply:
Done.